### PR TITLE
Fixed certificate issues on Operator restarts

### LINKED ISSUE #1694: Operator should store validator certificate in secret and not recreate it on every restart
We should be only refreshing certificate when its expiration date draws close, as every time certificate is recreated, controller must be restarted to reload its certificate used by validators and mutating webhooks.

### PR DESCRIPTION
**Pull request by hdkshingala**
### Description of change

This change persists the certificates generated by the Operator for webhooks and Controller so that we don't see the certificate errors during upgrades or operator pod restarts.

Fixes #1694

##### Checklist

- [x] Tested in playground or other setup
- [x] Documentation is changed or added

### Summary by OpenAI

**Bug fix:**
- Persist certificates generated by the Operator for webhooks and Controller to prevent certificate issues on restarts

**Documentation:**
- Update installation documentation to reflect certificate persistence changes

**Refactor:**
- Improve code readability and maintainability

> 🎉 Certificates persist, no more errors resist! 🚀
> With docs updated, and code refactored,
> Our Operator's journey, now unencumbered. 🌟

### CODE REVIEW

File: operator/controllers/mutate_func.go [hunk 1]
```diff
@@ -91,7 +96,6 @@ func ValidatingWebhookConfigurationMutate(vwc *admissionregistrationv1.Validatin
 	return func() error {
 		vwc.Webhooks[0].AdmissionReviewVersions = webhooks[0].AdmissionReviewVersions
 		vwc.Webhooks[0].ClientConfig = webhooks[0].ClientConfig
-		vwc.Webhooks[0].NamespaceSelector = webhooks[0].NamespaceSelector
 		vwc.Webhooks[0].ObjectSelector = webhooks[0].ObjectSelector
 		vwc.Webhooks[0].Rules = webhooks[0].Rules
 		vwc.Webhooks[0].FailurePolicy = webhooks[0].FailurePolicy
```
[the remark this answers is not in the document]
Comment (hdkshingala) [reply operator/controllers/mutate_func.go:101]: This is removed on purpose as it is not being used

File: operator/controllers/mutate_func.go [hunk 2]
```diff
@@ -62,9 +64,12 @@ func MutatingWebhookConfigurationMutate(mwc *admissionregistrationv1.MutatingWeb
...
-func SecretMutate(secret *corev1.Secret, data map[string][]byte) controllerutil.MutateFn {
+func SecretMutate(secret *corev1.Secret, data map[string][]byte, ownerReferences []v1.OwnerReference) controllerutil.MutateFn {
 	return func() error {
 		secret.Data = data
+		if !slices.Equal(secret.OwnerReferences, ownerReferences) {
+			secret.OwnerReferences = ownerReferences
+		}
 		return nil
 	}
 }
```
[the remark this answers is not in the document]
Comment (hdkshingala) [reply operator/controllers/mutate_func.go:75]: Usage of this library is fine

File: operator/controllers/utils.go [hunk 3]
```diff
@@ -607,6 +616,35 @@ func CheckAndGenerateCertForOperator() error {
...
+		}
+		_, err = controllerutil.CreateOrUpdate(context.Background(), k8sClient, secret, SecretMutate(secret, secret.Data, secret.OwnerReferences))
+		if err != nil {
+			return err
+		}
+	}
+
+	block, _ := pem.Decode(secret.Data[OperatorCertName])
+	cert, err := x509.ParseCertificate(block.Bytes)
+	if err == nil && cert.NotAfter.After(time.Now()) {
```
Comment (kklimonda-fn) [inline operator/controllers/utils.go:642]: What if err is not nil?

**Review by kklimonda-fn — APPROVED** [review]

File: operator/controllers/utils.go (hunk 3 above)
Comment (hdkshingala) [reply operator/controllers/utils.go:642]: If error is not nil, the new certificates which are generated above will be used instead

Comment (hdkshingala) [reply operator/controllers/utils.go:642]: If error is not nil, the new certificates which are generated above will be used instead

Comment (hdkshingala) [reply operator/controllers/utils.go:642]: I see a potential issue in this and fixing it. Thank you for pointing this out

File: operator/controllers/controller/secrets_test.go [hunk 4]
```diff
@@ -163,15 +163,6 @@ var _ = Describe("Secret for Controller Cert", func() {
...
-						{
-							APIVersion:         "fluxninja.com/v1alpha1",
-							Name:               instance.GetName(),
-							Kind:               "Controller",
-							Controller:         pointer.Bool(true),
-							BlockOwnerDeletion: pointer.Bool(true),
-						},
-					},
 				},
 				Data: map[string][]byte{
```
[the remark this answers is not in the document]
Comment (hdkshingala) [reply operator/controllers/controller/secrets_test.go:167]: This is done on purpose as we want to keep the resource even after deletion

File: operator/controllers/controller/secrets_test.go [hunk 5]
```diff
@@ -211,15 +202,6 @@ var _ = Describe("Secret for Controller Cert", func() {
...
```
[the remark this answers is not in the document]
Comment (hdkshingala) [reply operator/controllers/controller/secrets_test.go:206]: This is done on purpose as we want to keep the resource even after deletion

File: operator/controllers/utils.go [hunk 6]
```diff
@@ -607,6 +616,54 @@ func CheckAndGenerateCertForOperator() error {
...
+
+		_, err = controllerutil.CreateOrUpdate(context.Background(), k8sClient, secret, SecretMutate(secret, secret.Data, secret.OwnerReferences))
+		if err != nil {
+			return err
+		}
+	}
+
 	err = os.MkdirAll(os.Getenv("APERTURE_OPERATOR_CERT_DIR"), 0o777)
 	if err != nil {
 		return err
```
[the remark this answers is not in the document]
Comment (hdkshingala) [reply operator/controllers/utils.go:669]: New certificate is generate at line no 614

File: operator/controllers/namespace/reconciler.go [hunk 7]
```diff
@@ -148,16 +148,17 @@ func (r *NamespaceReconciler) reconcileControllerCertConfigMap(ctx context.Conte
...
 
+	if instance.Spec.ControllerClientCertConfig.ClientCertKeyName == "" {
+		instance.Spec.ControllerClientCertConfig.ClientCertKeyName = controllers.ControllerClientCertKey
+	}
+
 	configMap := agent.CreateAgentControllerClientCertConfigMapInNamespace(ctx, r.Client, instance, namespace)
 
 	if configMap == nil {
 		return nil
 	}
```
[the remark this answers is not in the document]
Comment (hdkshingala) [reply operator/controllers/namespace/reconciler.go:159]: ClientCertKeyName is needed in ConfigMap preparation

**Pull request merged**